### ISSUE TITLE
Mercurial repository type doesn't allow for SSH URLs

### ISSUE THREAD

**Issue by gaxweb** [composer/composer#11874]
We would like to use https://www.codebasehq.com/ to deploy our private Mercurial repositories via SSH and I've run into an issue here. If I use the `hg` repository type in the `composer.json`, I get the error:

```
In Hg.php line 90:
  Failed to clone ssh://hg@codebasehq.com/gaxweb/projectname/repository.hg,
  The given URL (…) does not match the required format (http(s)://(username:password@)example.com/path-to-repository)  
```   

HG URLs can start with "ssh" in general, so why is the `hg` type checking, if it starts with `http`? I can connect just fine, if I use the `vcs` type in the `composer.json` (but I have to enter the ssh key password 3 times) after all. So it's weird that it works one way but not the other.

Is this intentional or just an oversight? I could try my hand at a pull request. I think the problem lies here:

https://github.com/composer/composer/blob/66acb84c12d1d2b3cce7b2967f8e25522f502e60/src/Composer/Util/Hg.php#L61

**Comment by stof** [comment]
Mercurial repositories are much less used in Composer than Git repositories. So it would not surprise me if the support is incomplete (I suspect that support for Mercurial was mostly tested with Bitbucket in the past, and they supported the `https` scheme so this was not an issue).

**Comment by gaxweb** [comment]
CodebaseHQ does support them, too, but the way repository user permissions are implemented means it makes more sense to use SSH for private repositories.

It's not a pressing issue because the `vcs` type works, as I said. But it looks like an omission to me, so I created this issue.

**Comment by stof** [comment]
Feel free to create a PR fixing the issue.

**Comment by Seldaek** [comment]
Fixed by #11878
